static/js/components/Editor.js

```diff
@@ -0,0 +1,310 @@
+        this.view.focus()
+      }
+    }
+  }
+
+  dispatchTransaction(transaction: Object) {
+    // we need our own handler for this so we can hook into
+    // the EditorState lifecycle to synchronize it with the
+    // redux store (via the onChange prop)
+    const { onChange } = this.props
+
+    const state = this.view.state.apply(transaction)
+    this.view.updateState(state)
+    this.setState({ editorState: state })
+
+    onChange(defaultMarkdownSerializer.serialize(this.view.state.doc))
+  }
+
+  menuClickHandlerCreator = (command: Function) => (e: Object) => {
+    e.preventDefault()
+    command(this.view.state, this.view.dispatch, this.view)
+    this.view.focus()
+  }
+
+  renderMenuBar = () => {
+    const selectEmpty = this.selectionIsEmpty()
+
+    const linkIcon =
+      this.view && this.markIsActive(schema.marks.link) ? "link_off" : "link"
+
+    return (
+      <div className="menu-bar">
+        {menuBarManifest.map((buttonGroup, idx) => (
+          <div className="button-group" key={idx}>
+            {buttonGroup.map(({ command, title, icon, mark }) => (
+              <div
+                className={`menu-button ${
+                  mark && this.markIsActive(mark) ? "active" : ""
+                }`}
+                onClick={this.menuClickHandlerCreator(command)}
+                key={title}
+              >
+                {icon}
+              </div>
+            ))}
+          </div>
+        ))}
+        <div className="button-group">
+          <div
+            className={`menu-button ${selectEmpty ? "disabled" : ""}`}
+            onClick={selectEmpty ? null : this.handleLinkClick}
+          >
+            <i className={`material-icons ${linkIcon}`}>{linkIcon}</i>
+          </div>
+        </div>
+      </div>
+    )
+  }
+
+  markIsActive = (mark: Object) => {
+    // on the first render call this.view isn't there yet
+    if (!this.view) {
+      return false
+    }
+
+    const { from, $from, to, empty } = this.view.state.selection
+    return empty
+      ? !!mark.isInSet(this.view.state.storedMarks || $from.marks())
+      : this.view.state.doc.rangeHasMark(from, to, mark)
+  }
+
+  selectionIsEmpty = () => {
+    if (!this.view) {
+      return true
+    }
+
+    return (
+      this.view.state.selection.empty ||
+      this.view.state.selection
+        .content()
+        .content.content.map(node => node.textContent)
+        .join("")
+        .trim() === ""
+    )
+  }
+
+  handleLinkClick = () => {
+    if (this.markIsActive(schema.marks.link)) {
+      toggleMark(schema.marks.link)(this.view.state, this.view.dispatch)
+    } else {
+      this.openLinkMenu()
+    }
+  }
+
+  openLinkMenu = () => {
+    const { dispatch } = this.props
+    const { formBeginEdit } = this.formHelpers
+
+    this.setState({ showLinkMenu: true })
+    dispatch(formBeginEdit())
+  }
+
+  closeLinkMenu = () => {
+    this.setState({ showLinkMenu: false })
+  }
+
+  updateLinkMenuForm = (e: Object) => {
```

Comment:
Is `e` not of the type `Event`?

Reply:
I had typed this as `Object` because it gives me an error about not finding the `name` and `value` props in `target` if I set it to have type `Event`

Reply:
ah, but it works if I do `SyntheticInputEvent`! huzzah